File: DeepBackend/src/main/java/com/deepelectrocodingbackend/deepbackend/service/RecuService.java
package com.deepelectrocodingbackend.deepbackend.service;

import com.deepelectrocodingbackend.deepbackend.entity.Recu;
import com.deepelectrocodingbackend.deepbackend.repository.RecuRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.List;
import java.util.Optional;

@Service
public class RecuService {

    @Autowired
    private RecuRepository recuRepository;

    // Créer un nouveau reçu
    public Recu createRecu(Recu recu) {
        return recuRepository.save(recu);
    }

    // Récupérer tous les reçus
    public List<Recu> getAllRecus() {
        return recuRepository.findAll();
    }

    // Récupérer un reçu par ID
    public Optional<Recu> getRecuById(Long id) {
        return recuRepository.findById(id);
    }

    // Mettre à jour un reçu
    public Recu updateRecu(Long id, Recu recuDetails) {
        Recu recu = recuRepository.findById(id).orElseThrow(() -> new RuntimeException("Recu not found"));
        recu.setMontant(recuDetails.getMontant());
        recu.setDate(recuDetails.getDate());
        recu.setDescription(recuDetails.getDescription());
        return recuRepository.save(recu);
    }

    // Supprimer un reçu
    public void deleteRecu(Long id) {
        recuRepository.deleteById(id);
    }

    // Récupérer les reçus par montant
    public List<Recu> getRecusByMontant(double montant) {
        return recuRepository.findByMontant(montant);
    }

    // Récupérer les reçus par description
    public List<Recu> getRecusByDescription(String description) {
        return recuRepository.findByDescriptionContaining(description);
    }

    // Récupérer les reçus par date
    public List<Recu> getRecusByDateBetween(Date startDate, Date endDate) {
        return recuRepository.findByDateBetween(startDate, endDate);
    }
}

File: DeepFrontend/src/app/services/recu.service.ts
// src/app/services/recu.service.ts

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Recu } from '../models/recu.model';

@Injectable({
  providedIn: 'root'
})
export class RecuService {
  private apiUrl = 'http://localhost:8080/api/recus'; // Remplacez par l'URL de votre API

  constructor(private http: HttpClient) {}

  // Créer un nouveau reçu
  createRecu(recu: Recu): Observable<Recu> {
    return this.http.post<Recu>(this.apiUrl, recu);
  }

  // Récupérer tous les reçus
  getAllRecus(): Observable<Recu[]> {
    return this.http.get<Recu[]>(this.apiUrl);
  }

  // Récupérer un reçu par ID
  getRecuById(id: number): Observable<Recu> {
    return this.http.get<Recu>(`${this.apiUrl}/${id}`);
  }

  // Mettre à jour un reçu
  updateRecu(id: number, recu: Recu): Observable<Recu> {
    return this.http.put<Recu>(`${this.apiUrl}/${id}`, recu);
  }

  // Supprimer un reçu
  deleteRecu(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Récupérer les reçus par montant
  getRecusByMontant(montant: number): Observable<Recu[]> {
    return this.http.get<Recu[]>(`${this.apiUrl}/montant/${montant}`);
  }

  // Récupérer les reçus par description
  getRecusByDescription(description: string): Observable<Recu[]> {
    return this.http.get<Recu[]>(`${this.apiUrl}/description/${description}`);
  }

  // Récupérer les reçus par date
  getRecusByDateBetween(startDate: Date, endDate: Date): Observable<Recu[]> {
    const params = { startDate: startDate.toISOString(), endDate: endDate.toISOString() };
    return this.http.get<Recu[]>(`${this.apiUrl}/date`, { params });
  }
}